source/common/tcp_proxy/hash_policy.cc

```diff
@@ -0,0 +1,54 @@
+#include "common/tcp_proxy/hash_policy.h"
+#include "envoy/common/exception.h"
+
+namespace Envoy {
+namespace TcpProxy {
+
+class SourceIpHashMethod : public HashPolicyImpl::HashMethod {
+public:
+  absl::optional<uint64_t> evaluate(const Network::Address::Instance* downstream_addr,
+                                    const Network::Address::Instance*) const override {
+    if (downstream_addr && downstream_addr->ip() &&
+        !downstream_addr->ip()->addressAsString().empty()) {
+      return HashUtil::xxHash64(downstream_addr->ip()->addressAsString());
+    }
+
+    return absl::nullopt;
+  }
+};
+
+HashPolicyImpl::HashPolicyImpl(
+    absl::Span<const envoy::config::filter::network::tcp_proxy::v2::TcpProxy::HashPolicy* const>
+        hash_policies) {
+  hash_impls_.reserve(hash_policies.size());
+
+  for (auto* hash_policy : hash_policies) {
+    switch (hash_policy->policy_specifier_case()) {
+    case envoy::config::filter::network::tcp_proxy::v2::TcpProxy::HashPolicy::kSourceIp:
+      hash_impls_.emplace_back(new SourceIpHashMethod());
+      break;
+    default:
+      throw EnvoyException(
```

Comment:
This is checked by schema and can't happen. I would use NOT_REACHED here.

Reply:
Done